### ISSUE TITLE
Proposal: sort set values from command line before merging

### ISSUE THREAD

**Issue by fengxsong** [helm/helm#4503]
## test1 - A universal and inherited chart

values.yaml

```
...
ports: []
...
``` 

service.yaml

```
ports:
{{- range .Values.ports }}
- port: {{ .port }}
  protocol: {{ .protocol }}
{{- end }}
```

## chart test2

requirements.yaml

```
dependencies:
- name: test1
...
```

if I want to override the ports in chart test2

```
/ $: helm install --set test1.ports\[0\].port=8080 --set test1.ports\[0\].protocol=TCP --set test1.ports\[1\].port=9090 --set test1.ports\[1\].protocol=TCP ...
```
this command works as expected, but if I change the order of the  sets(In most cases, it is impossible to do so), like

```
/ $: helm install --set test1.ports\[1\].port=8080 --set test1.ports\[0\].protocol=TCP --set test1.ports\[0\].port=9090 --set test1.ports\[1\].protocol=TCP ...
```

This will lead to assertion panic

```
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 1 [running]:
k8s.io/helm/pkg/strvals.(*parser).listItem(0xc42087bf80, 0xc42087bf40, 0x2, 0x2, 0x0, 0x1, 0x5b, 0x0, 0x0, 0x0)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:291 +0xef3
k8s.io/helm/pkg/strvals.(*parser).key(0xc42087bf80, 0xc420807530, 0xc420cdd4b8, 0x3)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:177 +0xc1b
k8s.io/helm/pkg/strvals.(*parser).key(0xc42087bf80, 0xc420761f80, 0xc420cdd6f0, 0x8)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:215 +0x1a3
k8s.io/helm/pkg/strvals.(*parser).key(0xc42087bf80, 0xc420761f20, 0x1e14d7d, 0x20d7200)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:215 +0x1a3
k8s.io/helm/pkg/strvals.(*parser).parse(0xc42087bf80, 0xc420761f20, 0x0)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:133 +0x38
k8s.io/helm/pkg/strvals.ParseInto(0x7ffeefbff8c2, 0x22, 0xc420761f20, 0x0, 0x0)
	/go/src/k8s.io/helm/pkg/strvals/parser.go:85 +0xbf
main.vals(0x0, 0x0, 0x0, 0xc420274600, 0xd, 0x10, 0x2cd7b28, 0x0, 0x0, 0x2cd7b28, ...)
	/go/src/k8s.io/helm/cmd/helm/install.go:382 +0x45a
main.(*installCmd).run(0xc4201d0600, 0x0, 0x2378600)
	/go/src/k8s.io/helm/cmd/helm/install.go:232 +0x1e4
main.newInstallCmd.func2(0xc4207666c0, 0xc420310780, 0x1, 0x1d, 0x0, 0x0)
	/go/src/k8s.io/helm/cmd/helm/install.go:191 +0x326
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).execute(0xc4207666c0, 0xc4203101e0, 0x1d, 0x1e, 0xc4207666c0, 0xc4203101e0)
	/go/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:599 +0x3db
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4207b0000, 0xc42071ab40, 0xc42071b440, 0xc42071bd40)
	/go/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:689 +0x2d4
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4207b0000, 0x1e, 0x1f)
	/go/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.main()
	/go/src/k8s.io/helm/cmd/helm/helm.go:161 +0x79
```

Sort values before invoking function `strvals.ParseInto` and `strvals.ParseIntoString` might do the trick

**Comment by spoortiprabhu** [comment]
@fengxsong can you please tell me how did you added ports? where does that ports get added?

**Comment by fengxsong** [comment]
@spoortiprabhu I haved updated the issue.

**Comment by spoortiprabhu** [comment]
@fengxsong ports are not getting to that chart even if we run the command "$: helm install --set test1.ports[0].port=8080 --set test1.ports[0].protocol=TCP --set test1.ports[1].port=9090 --set test1.ports[1].protocol=TCP ...".

**Comment by fengxsong** [comment]
@spoortiprabhu click [here](https://gist.github.com/fengxsong/03c667d204b156c62fdc64efa8441f65) for more details

**Comment by danjwinter** [comment]
Hello! In the last few days some colleagues and I decided to use some hackathon time to work on this. We managed to make a failing test case and then modified some code in the strvals parser that made that test case pass.

Instead of sorting values, our solution handles a panic that happens when a type assertion of list[i] to map[string]interface{} fails. Our [commit](https://github.com/danjwinter/helm/commit/3656d4bba3235528268be19f05a07cddae8f9259).

We were getting ready to make a PR, but then noticed something that we thought might merit some discussion. 

With or without our change, if there is a list of ports represented in a values.yaml, using the `--set` flag on one item in the list overwrites the entire list instead of just the item we referenced via a particular index. This implies that one would have to call `--set` for each item in the list to overwrite the item value that one cares about. It looks like this behavior occurs in the code here:

https://github.com/helm/helm/blob/8be42bae885a04b4acc242cf420911145b32ee1c/cmd/helm/install.go#L337-L342

Is this expected? If so, we'll continue with a PR. If not, we'll look at modifying our solution.
